### PR TITLE
Support block gap created by fast sync

### PR DESCRIPTION
**Pull request by liuchengxu**
This is part 2 of https://github.com/paritytech/polkadot-sdk/issues/5406#issuecomment-2325064863, properly handling the block gap generated during fast sync.

Although #5406 remains unresolved due to the known issues in #5663, I decided to open up this PR earlier than later to speed up the overall progress. I've tested the fast sync locally with this PR, and it appears to be functioning well. (I was doing a fast sync from a discontinued archive node locally, thus the issue highlighted in https://github.com/paritytech/polkadot-sdk/pull/5663#discussion_r1752124039 was bypassed exactly.) 

Once the edge cases in #5663 are addressed, we can move forward by removing the body attribute from the LightState block request and complete the work on #5406. The changes in this PR are incremental, so reviewing commit by commit should provide the best clarity.

cc @dmitry-markin

### CODE REVIEW

File: substrate/client/db/src/lib.rs [hunk 1]
```diff
@@ -1486,6 +1486,7 @@ impl<Block: BlockT> Backend<Block> {
 				.map(|(n, _)| n)
 				.unwrap_or(Zero::zero());
 			let existing_header = number <= highest_leaf && self.blockchain.header(hash)?.is_some();
+			let body_exists = pending_block.body.is_some();
```
Comment (dmitry-markin) [inline substrate/client/db/src/lib.rs:1489]: Nit: to match `existing_header` naming above.
```suggestion
			let existing_body = pending_block.body.is_some();
```

**Review by dmitry-markin — APPROVED** [review]
Overall looks good to me, but do you think we should support converting already present "missing body" gap into "missing header and body" gap if detect it?

**Comment by liuchengxu** [comment]
I don't believe supporting this conversion is necessary unless we encounter a specific use case that justifies it. Currently, the conceptual model is straightforward: gaps are caused either by warp sync or fast sync, but not both simultaneously. From my point of view, warp sync is typically the preferred method, with fast sync used only when warp sync is unsupported by the chain. Users are generally expected to choose one sync strategy, rather than a mix of the two. Adding support for converting gaps between the two could introduce unnecessary complexity without clear benefits.

**Comment by liuchengxu** [comment]
Ping @bkchr In addition, some insights in terms of the edge case in https://github.com/paritytech/polkadot-sdk/pull/5663#discussion_r1752124039 would also be appreciated.

**Comment by liuchengxu** [comment]
Could you have a look at this? @bkchr

**Review by bkchr — APPROVED** [review]

File: prdoc/pr_5703.prdoc [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  - audience: Node Dev
+    description: |
+      Implements support for handling block gaps generated during fast sync. This includes managing the creation,
+      updating, and removal of block gaps.
+      Note that this feature is not fully activated until the `body` attribute is removed from the `LightState`
+      block request in chain sync, which will occur after the issue #5406 is resolved.
+
+crates:
+  - name: sc-client-db
+    bump: none
```
Comment (bkchr) [inline prdoc/pr_5703.prdoc:13]:
```suggestion
    bump: patch
```

**Pull request merged**